fun.lua

```diff
@@ -1032,6 +1040,86 @@ methods.operator = operator
 exports.op = operator
 methods.op = operator
 
+--------------------------------------------------------------------------------
+-- Stateful iterators
+--------------------------------------------------------------------------------
+
+-- Returns 'gen' throwing an error that explains that method is not supported
+local unsupported_stateful_method_gen = function(method)
+    return function()
+        error(string.format("stateful iterators do not support method '%s'",
+              method))
+    end
+end
+
+-- Helper of 'wrap_with_elems'
+local gen_with_elems_wrapper = function(gen_param_elems, state)
+    local elems = gen_param_elems[3]
+    if not elems then
+        return gen_param_elems[1](gen_param_elems[2], state)
+    else
+        gen_param_elems[3] = nil
+        return state, unpack(elems)
+    end
+end
+
+-- Wrap an iterator that yields all values from 'elems' table on the first
+-- iteration and then works in the same way as 'wrap(gen, param, state)'
+local wrap_with_elems = function(elems, gen, param, state)
+    local gen_param_elems = {gen, param, elems}
+    return wrap(gen_with_elems_wrapper, gen_param_elems, state)
+end
+
+-- Checks if drop_while should continue skipping. If iterator is not exhausted,
+-- elements returned by iterator are wrapped into a table and returned as the
+-- third return value
+local stateful_drop_while_check = function(fun, state_x, ...)
+    if state_x == nil then
+        return state_x, false
+    end
+    return state_x, fun(...), {...}
+end
+
+-- Implementation of method drop_while: original one copies iterator state to
+-- iterate over the same element twice - it doesn't work with stateful iterators
+local stateful_drop_while = function(fun, gen_x, param_x, state_x)
+    assert(type(fun) == "function", "invalid first argument to drop_while")
+    local cont = true
+    local elems
+    while state_x ~= nil and cont do
+        state_x, cont, elems = stateful_drop_while_check(fun, gen_x(param_x, state_x))
+    end
+    if state_x == nil then
+        return wrap(nil_gen, nil, nil)
+    end
+    return wrap_with_elems(elems, gen_x, param_x, state_x)
+end
+
+local stateful_methods = deepcopy(methods)
+-- Method 'is_null' cannot be supported since it advances iterator and
+-- does not return any new iterator, so passed stateful iterator is advanced
+-- and returned value is lost
+stateful_methods.is_null = unsupported_stateful_method_gen('is_null')
```

Comment:
I would like to share some raw ideas around problems we met with stateful iterators.

Observation: we're limited by the API that Lua provides for iterators. The `gen` function returns the next value (or values) and advances the iterator. There is no API to define a lookahead operation.

OTOH, luafun can provide an optional API to allow a user to define such a function. For example, if `gen` is a callable table if may have extra functions as fields:

```lua
local function myiterator_new()
    return fun.wrap(setmetatable({
        gen = myiterator_gen,
        lookahead = myiterator_lookahead,
        advance = myiterator_advance,
        stepback = myiterator_stepback,
        rewind = myiterator_rewind,
    }, {
        __call = myiterator_gen,
    }), myparam, mystate)
end
```

(The set of the extra functions and their names are just for example.)

This way luafun can use extra functions if they're provided: for example, `is_null` can use `lookahead`. So, a stateful iterator still may be handled correctly by `is_null` if it provides this additional function.

Another idea is to spins around an analogue of `ungetc` or a kind of bufferized iterators. OTOH, it should be supported in all the functions to have a chance to work correctly and it would significantly increase complexity of luafun's code.

However, we can provide a primitive to create a bufferized iterator and push the received element back in `is_null` if the iterator supports it.

For example (not tested, just to illustrate the idea),

```lua
-- user's code.

local function myiterator_new()
    return fun.bufferize(myiterator_gen, myparam, mystate)
end

-- luafun's code

-- TODO: Support multiple values.
local function pushback(self, v)
    table.insert(self.buf, v)
    return v
end

local function bufferized_gen(param, state)
    local orig_gen = param[1]
    local orig_param = param[2]
    local buf = param[3]
    if next(buf) ~= nil then
        return state, table.remove(buf)
    end
    return orig_gen(orig_param, state)
end

function fun.bufferize(gen, param, state)
    local buf = {}
    return fun.wrap(setmetatable({
        buf = buf,
        pushback = pushback,
    }, {
        __call = bufferized_gen,
    }), {gen, param, buf}, state)
end

local function has_pushback(gen)
    return type(gen) == 'table' and gen.pushback ~= nil
end

local function is_null(gen, param, state)
    if has_pushback(gen) then
        return gen.pushback(gen(param, state)) == nil
    end
    <..existing implementation..>
    return gen(param, deepcopy(state)) == nil
end
```

This way a user also can implement its own `pushback` function without using `fun.bufferize()`.

However, an iterator created from the bufferized one (say, using `:map()`) can't be detected as bufferized. So, where `it:is_null()` works, `it:map(<...>):is_null()` doesn't.

----

Within this patchset I would leave `is_null` forbidden if we're going to provide the separate `iter_stateful` and `wrap_stateful` constructors or leave it unchanged if we will decide to don't introduce them. This comment is just to share some thoughts how we can solve the given problems in a future if we see a need to work on it.

Reply:
That's good idea. But now it seems to be overkill, especially when we decided to simply re-implement `drop_while`.